util/collate/collate.go

```diff
@@ -111,62 +125,79 @@ func GetCollator(collate string) Collator {
 	if atomic.LoadInt32(&newCollationEnabled) == 1 {
 		ctor, ok := newCollatorMap[collate]
 		if !ok {
+			logutil.BgLogger().Warn(
+				"Unable to get collator by name, use binCollator instead.",
+				zap.String("name", collate),
+				zap.Stack("stack"))
 			return newCollatorMap["utf8mb4_bin"]
 		}
 		return ctor
 	}
-	ctor, ok := collatorMap[collate]
-	if !ok {
-		return collatorMap["utf8mb4_bin"]
-	}
-	return ctor
+	return binCollatorInstance
 }
 
 // GetCollatorByID get the collator according to id, it will return the binary collator if the corresponding collator doesn't exist.
 func GetCollatorByID(id int) Collator {
 	if atomic.LoadInt32(&newCollationEnabled) == 1 {
 		ctor, ok := newCollatorIDMap[id]
 		if !ok {
+			logutil.BgLogger().Warn(
+				"Unable to get collator by ID, use binCollator instead.",
+				zap.Int("ID", id),
+				zap.Stack("stack"))
 			return newCollatorMap["utf8mb4_bin"]
 		}
 		return ctor
 	}
-	ctor, ok := collatorIDMap[id]
-	if !ok {
-		return collatorMap["utf8mb4_bin"]
-	}
-	return ctor
-}
-
-type binCollator struct {
-}
-
-// Compare implement Collator interface.
-func (bc *binCollator) Compare(a, b string) int {
-	return strings.Compare(a, b)
-}
-
-// Key implement Collator interface.
-func (bc *binCollator) Key(str string) []byte {
-	return []byte(str)
-}
-
-// Pattern implements Collator interface.
-func (bc *binCollator) Pattern() WildcardPattern {
-	return &binPattern{}
+	return binCollatorInstance
 }
 
 // CollationID2Name return the collation name by the given id.
 // If the id is not found in the map, the default collation is returned.
 func CollationID2Name(id int32) string {
 	name, ok := mysql.Collations[uint8(id)]
 	if !ok {
+		// TODO(bb7133): fix repeating logs when the following code is uncommented.
+		//logutil.BgLogger().Warn(
+		//	"Unable to get collation name from ID, use default collation instead.",
+		//	zap.Int32("ID", id),
+		//	zap.Stack("stack"))
 		return mysql.DefaultCollationName
 	}
 	return name
 }
 
-type binPaddingCollator struct {
+// GetCollationByName wraps charset.GetCollationByName, it checks the collation.
+func GetCollationByName(name string) (coll *charset.Collation, err error) {
+	if coll, err = charset.GetCollationByName(name); err != nil {
+		return nil, errors.Trace(err)
+	}
+	if atomic.LoadInt32(&newCollationEnabled) == 1 {
```

Comment:
```suggestion
	if NewCollationEnabled() {
```

Same for all the `atomic.LoadInt32(&newCollationEnabled) == 1` in this file.

Reply:
Got your point, but this code style is commonly used by collate.go, I think we can get another PR for them all.